### PR TITLE
[API] sorting product by position in taxon

### PR DESCRIPTION
**Pull request by AdamKasp**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/SortProductsByPositionInTaxon.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        }
+
+        $user = $this->userContext->getUser();
+        if ($user !== null && in_array('ROLE_API_ACCESS', $user->getRoles(), true)) {
+            return;
+        }
+
+        $rootAlias = $queryBuilder->getRootAliases()[0];
+        $queryBuilder
+            ->distinct()
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/SortProductsByPositionInTaxon.php:50]: Why have you decided for distinct here?

File: src/Sylius/Behat/Context/Api/Shop/ProductContext.php [hunk 2]
```diff
@@ -225,6 +226,28 @@ public function iShouldSeeEmptyListOfProducts(): void
         Assert::same($this->responseChecker->countTotalCollectionItems($this->client->getLastResponse()), 0);
     }
 
+    /**
+     * @Then I should see :count products in the list
+     */
+    public function iShouldSeeProductsInTheList(int $count): void
+    {
+        Assert::eq($this->responseChecker->countCollectionItems($this->client->getLastResponse()), $count);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/ProductContext.php:234]:
```suggestion
        Assert::same($this->responseChecker->countCollectionItems($this->client->getLastResponse()), $count);
```

File: src/Sylius/Behat/Context/Api/Shop/ProductContext.php [hunk 3]
```diff
@@ -225,6 +226,28 @@ public function iShouldSeeEmptyListOfProducts(): void
...
+     */
+    public function theyShouldHaveOrderLikeAnd(string ...$productNames): void
+    {
+        $productNamesFromResponse = new ArrayCollection();
+
+        foreach ($this->responseChecker->getCollection($this->client->getLastResponse()) as $productItem) {
+            $productNamesFromResponse->add($productItem['translations']['en_US']['name']);
+        }
+
+        Assert::allInArray($productNamesFromResponse, $productNames);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/ProductContext.php:248]: Are you sure this method checks the order of elements?

File: src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/SortProductsByPositionInTaxonSpec.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+        $queryBuilder
+            ->innerJoin('o.productTaxons', 'productTaxon')
+            ->willReturn($queryBuilder)
+        ;
+
+        $queryBuilder
+            ->addOrderBy('productTaxon.position')
+            ->willReturn($queryBuilder)
+        ;
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/SortProductsByPositionInTaxonSpec.php:91]:
```suggestion
        $queryBuilder->getRootAliases()->willReturn(['o']);
        $queryBuilder->distinct()->willReturn($queryBuilder);
        $queryBuilder->addSelect('productTaxon')->willReturn($queryBuilder);
        $queryBuilder->innerJoin('o.productTaxons', 'productTaxon')->willReturn($queryBuilder);
        $queryBuilder->addOrderBy('productTaxon.position')->willReturn($queryBuilder);
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, Adam! :1st_place_medal:

**Pull request merged**